### PR TITLE
PIM-8305: Fix product count displayed by Mass delete job

### PR DESCRIPTION
**Pull request by pchasle**
**Description (for Contributor and Core Developer)**

Fix the **counter** for number of _products_ and _product models_ deleted by the **Mass delete** process on the execution details screen.

When a mass delete of product models is launched, doctrine is automatically removing children (product models and products) with a cascade delete.
But the execution details screen only display the number of _products_ and _product models_ given to the job.

E.g. for a the _product model_ `Caelus` with 2 levels of variant:

With a sub-level `color` with 2 _product models_:

And each `color` have 3 `size`:

The amount of deleted _products_ and _product models_ deleted for `Caelus` should be:
- _Products_: 2 `colors` * 3 `sizes` = 6
- _Product models_: 1 root model + 2 `colors` = 3

Before:

After:

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | OK
| Added acceptance tests            | -
| Added integration tests           | OK
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+declare(strict_types=1);
+
+namespace Pim\Component\Catalog\ProductAndProductModel\Query;
+
+use Pim\Component\Catalog\Model\ProductModelInterface;
+
+/**
+ * @todo pull-up 3.x Move to `Akeneo\Pim\Enrichment\Component\Product\ProductAndProductModel\Query`
+ *
```
Comment (jmleroux) [inline src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php]: Could you please add a functional description?

File: src/Pim/Component/Catalog/ProductModel/Query/CountProductModelChildrenInterface.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+declare(strict_types=1);
+
+namespace Pim\Component\Catalog\ProductModel\Query;
+
+use Pim\Component\Catalog\Model\ProductModelInterface;
+
+/**
+ * @todo pull-up 3.x Move to `Akeneo\Pim\Enrichment\Component\Product\ProductModel\Query`
+ *
```
Comment (jmleroux) [inline src/Pim/Component/Catalog/ProductModel/Query/CountProductModelChildrenInterface.php]: Same, description please :wink:

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pim\Bundle\CatalogBundle\Doctrine\ORM\Query;
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: wrong namespace, it's sql query, not ORM

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+use Pim\Component\Catalog\Model\ProductModelInterface;
+use Pim\Component\Catalog\ProductModel\Query\CountProductModelChildrenInterface;
+
+/**
+ * @todo pull-up 3.x Move to `Akeneo\Pim\Enrichment\Bundle\Doctrine\ORM\Query`
+ *
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class CountProductModelChildren implements CountProductModelChildrenInterface
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: I don't like the naming as the parent product model is included

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+{
+    /** @var Connection */
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function forProductModels(array $productModels): int
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php]: forParentProductModels / productModelCodes

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /** @var Connection */
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function forProductModels(array $productModels): int
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: productModelCodes

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+            where level_1.id in (:productModelIds)
+SQL;
+
+        $stmt = $this->connection->executeQuery(
+            $sql,
+            ['productModelIds' => $productModelIds],
+            ['productModelIds' => Connection::PARAM_INT_ARRAY]
+        );
+
+        return (int)$stmt->fetchColumn();
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: bug if the list is empty or the product models do not exist

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            where level_1.id in (:productModelIds)
+SQL;
+
+        $stmt = $this->connection->executeQuery(
+            $sql,
+            ['productModelIds' => $productModelIds],
+            ['productModelIds' => Connection::PARAM_INT_ARRAY]
+        );
+
+        return (int)$stmt->fetchColumn();
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php]: same: bug if no product model

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pim\Bundle\CatalogBundle\Doctrine\ORM\Query;
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php]: wrong namespace

File: src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductModelChildrenIntegration.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                $repository->findOneByIdentifier('amor'),
+                $repository->findOneByIdentifier('caelus'),
+            ]
+        );
+        self::assertEquals($result, 4);
+    }
+
+    protected function getConfiguration(): Configuration
+    {
+        return $this->catalog->useFunctionalCatalog('catalog_modeling');
```
Comment (ahocquard) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductModelChildrenIntegration.php]: please avoid this catalog, it's very very long to load...

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php [hunk 11]
```diff
@@ -176,11 +184,31 @@ protected function doDelete(array $entities): void
         $products = $this->filterProducts($entities);
         $productModels = $this->filterProductModels($entities);
 
+        /* @todo pull-up 3.x Remove `if-else` */
+        if (null !== $this->countProductVariants) {
+            $deletedProducts = count($products);
```
Comment (ahocquard) [inline src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: `$deletedProducts = $this->countDeletedProduct();`


It would not pollute the function

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php [hunk 12]
```diff
@@ -176,11 +184,31 @@ protected function doDelete(array $entities): void
...
+            if (count($productModels) > 0) {
+                $deletedProducts += $this->countProductVariants->forProductModels($productModels);
```
Comment (ahocquard) [inline src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: your service should support empty array (becasue it's a valid array) and then:

```
$deletedProducts = null !== $this->countProductVariants ? 
    count($products) + $this->countProductVariants->forProductModels($productModels)
     : count($products);
```

(Ternary is  matter of style, as you wish)

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php [hunk 13]
```diff
@@ -176,11 +184,31 @@ protected function doDelete(array $entities): void
...
+            }
+        } else {
+            $deletedProducts = count($products); // to remove
+        }
+
+        /* @todo pull-up 3.x Remove `if-else` */
+        if (null !== $this->countProductModelChildren) {
```
Comment (ahocquard) [inline src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: same

File: src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface CountProductVariantsInterface
+{
+    /**
+     * @param ProductModelInterface[] $productModels
+     */
+    public function forProductModels(array $productModels): int;
```
Comment (ahocquard) [inline src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php]: productModelCodes, no need to pass the whole object. It prevents to use this query by not loading the ProductModelInterface otherwise.

File: src/Pim/Component/Catalog/ProductModel/Query/CountProductModelChildrenInterface.php [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface CountProductModelChildrenInterface
+{
+    /**
+     * @param ProductModelInterface[] $productModels
+     */
+    public function forProductModels(array $productModels): int;
```
Comment (ahocquard) [inline src/Pim/Component/Catalog/ProductModel/Query/CountProductModelChildrenInterface.php]: same

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php (hunk 3 above)
Comment (pchasle) [reply src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: I had a discussion about this with @jmleroux , but since it's in 2.3 the namespace didn't really exists and it will be moved to correct location while pulling-up.
Do you think we still need to change it ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php (hunk 6 above)
Comment (jmleroux) [reply src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: @ahocquard do not hesitate to use the "suggestion" feature of Github :wink:

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php (hunk 8 above)
Comment (pchasle) [reply src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php]: count will return 0 if there is no product

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php (hunk 4 above)
Comment (pchasle) [reply src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: I agree with you, it was hard to find a good name on the moment. Finally I changed it to `CountProductModelsAndChildrenProductModels`, it's a little long and repetitive but it's speak for itself. If it's fine by you ?

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php (hunk 5 above)
Comment (pchasle) [reply src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php]: With the new name `CountProductModelsAndChildrenProductModels` I think it's weird to have the notion of _parent_ in the name as the product models given in input are counted in the output.

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php (hunk 6 above)
Comment (pchasle) [reply src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelChildren.php]: Yes `code` is way better :+1:

File: src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php (hunk 11 above)
Comment (pchasle) [reply src/Pim/Component/Enrich/Job/DeleteProductsAndProductModelsTasklet.php]: Yes it's much clearer now

**Comment by pchasle** [comment]
I added way more exhaustive integration tests for the two queries and this allowed me to detect an edge case on `CountProductModelsAndChildrenProductModels.php` fixed with the help of @mmetayer

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelsAndChildrenProductModels.php [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+{
+    /** @var Connection */
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function forProductModels(array $productModelCodes): int
```
Comment (mmetayer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductModelsAndChildrenProductModels.php]: fromProductModelCodes

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+{
+    /** @var Connection */
+    private $connection;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function forProductModels(array $productModelCodes): int
```
Comment (mmetayer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php]: fromProductModelCodes

File: src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+use Doctrine\DBAL\Connection;
+use Pim\Component\Catalog\ProductAndProductModel\Query\CountProductVariantsInterface;
+
+/**
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @todo pull-up 3.x Move to `Akeneo\Pim\Enrichment\Bundle\Product\Query\Sql`
+ */
+final class CountProductVariants implements CountProductVariantsInterface
```
Comment (mmetayer) [inline src/Pim/Bundle/CatalogBundle/Doctrine/ORM/Query/CountProductVariants.php]: CountVariantProducts

File: src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductModelsAndChildrenProductModelsIntegration.php [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        // Duplicate of product models with 2 level of variant.
+        $result = $this->getQuery()->forProductModels(['a_shirt', 'a_shirt']);
+        self::assertEquals(3, $result);
+
+        // Level 2 product model and level 1 product model of the same product model with 2 levels of variant.
+        $result = $this->getQuery()->forProductModels(['a_small_shirt', 'a_shirt']);
+        self::assertEquals(3, $result);
+    }
+
+    public function setUp(): void
```
Comment (mmetayer) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductModelsAndChildrenProductModelsIntegration.php]: protected ;)

File: src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductVariantsIntegration.php [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+ *      - a_blue_shoes
+ *
+ * - a_shirt
+ *      - a_small_shirt
+ *          - a_red_small_shirt
+ *          - a_blue_small_shirt
+ *      - a_medium_shirt
+ *          - a_red_medium_shirt
+ */
+class CountProductVariantsIntegration extends TestCase
```
Comment (mmetayer) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductVariantsIntegration.php]: CountVariantProductsIntegration

File: src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductVariantsIntegration.php [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+        // Duplicate of product models with 2 level of variant.
+        $result = $this->getQuery()->forProductModels(['a_shirt', 'a_shirt']);
+        self::assertEquals(3, $result);
+
+        // Level 2 product model and level 1 product model of the same product model with 2 levels of variant.
+        $result = $this->getQuery()->forProductModels(['a_small_shirt', 'a_shirt']);
+        self::assertEquals(3, $result);
+    }
+
+    public function setUp(): void
```
Comment (mmetayer) [inline src/Pim/Bundle/CatalogBundle/tests/integration/Doctrine/ORM/Query/CountProductVariantsIntegration.php]: protected

File: src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php [hunk 22]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Pim\Component\Catalog\ProductAndProductModel\Query;
+
+/**
+ * Count the total number of product variants for a given list of product models.
```
Comment (mmetayer) [inline src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php]: variant products

File: src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @todo pull-up 3.x Move to `Akeneo\Pim\Enrichment\Component\Product\ProductAndProductModel\Query`
+ */
+interface CountProductVariantsInterface
```
Comment (mmetayer) [inline src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php]: CountVariantProductsInterface

File: src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+{
+    /**
+     * @param string[] $productModelCodes
+     */
+    public function forProductModels(array $productModelCodes): int;
```
Comment (mmetayer) [inline src/Pim/Component/Catalog/ProductAndProductModel/Query/CountProductVariantsInterface.php]: fromProductModelCodes

File: src/Pim/Component/Catalog/ProductModel/Query/CountProductModelsAndChildrenProductModelsInterface.php [hunk 25]
```diff
@@ -0,0 +1,21 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ *
+ * @todo pull-up 3.x Move to `Akeneo\Pim\Enrichment\Component\Product\ProductModel\Query`
+ */
+interface CountProductModelsAndChildrenProductModelsInterface
+{
+    /**
+     * @param string[] $productModelCodes
+     */
+    public function forProductModels(array $productModelCodes): int;
```
Comment (mmetayer) [inline src/Pim/Component/Catalog/ProductModel/Query/CountProductModelsAndChildrenProductModelsInterface.php]: fromProductModelCodes
